File: src/pipeline/metrics.ts
import type { Client } from "@libsql/client";
import type { Config } from "./config.js";
import { platformFilter, queryMessages } from "./db.js";
import { Platform, Role } from "./models.js";
import { POLITENESS_PATTERNS, BACKTRACK_PATTERNS, COMMAND_PATTERN, computeNlpMetrics } from "./nlp.js";
import { classifyPersona } from "./persona.js";
import { computeTrendMetrics } from "./trends.js";

function round(n: number, decimals: number): number {
  const f = 10 ** decimals;
  return Math.round(n * f) / f;
}

function countAllMatches(texts: string[], pattern: RegExp): number {
  let total = 0;
  for (const t of texts) {
    total += (t.match(pattern) ?? []).length;
    pattern.lastIndex = 0;
  }
  return total;
}

export async function computeModelUsage(
  client: Client,
  platform?: Platform,
): Promise<Record<string, any>> {
  const pf = platformFilter(platform);

  const totalHuman = Number((await client.execute({ sql: `SELECT COUNT(*) as cnt FROM messages WHERE role = 'human'${pf.clause}`, args: pf.args })).rows[0].cnt);
  const withMeta = Number((await client.execute({ sql: `SELECT COUNT(*) as cnt FROM messages WHERE role = 'human' AND (model_id IS NOT NULL OR model_provider IS NOT NULL)${pf.clause}`, args: pf.args })).rows[0].cnt);
  const coveragePct = totalHuman ? round((withMeta / totalHuman) * 100, 1) : 0;

  // By model with source breakdown
  const modelSourceRows = (await client.execute({
    sql: `SELECT COALESCE(model_id, 'unknown'), COALESCE(model_provider, 'unknown'), platform, COUNT(*) as cnt, SUM(word_count) as words FROM messages WHERE role = 'human' AND (model_id IS NOT NULL OR model_provider IS NOT NULL)${pf.clause} GROUP BY 1, 2, platform`,
    args: pf.args,
  })).rows;

  const byModelMap = new Map<string, any>();
  for (const r of modelSourceRows) {
    const key = `${r[0]}|${r[1]}`;
    if (!byModelMap.has(key)) byModelMap.set(key, { model_id: String(r[0]), model_provider: String(r[1]), prompts: 0, words: 0, sources: {} });
    const entry = byModelMap.get(key)!;
    entry.prompts += Number(r.cnt ?? r[3]);
    entry.words += Number(r.words ?? r[4]);
    entry.sources[String(r.platform ?? r[2])] = Number(r.cnt ?? r[3]);
  }

  const convRows = (await client.execute({
    sql: `SELECT COALESCE(model_id, 'unknown'), COALESCE(model_provider, 'unknown'), COUNT(DISTINCT conversation_id) as cnt FROM messages WHERE role = 'human' AND (model_id IS NOT NULL OR model_provider IS NOT NULL)${pf.clause} GROUP BY 1, 2`,
    args: pf.args,
  })).rows;
  const convMap = new Map(convRows.map((r) => [`${r[0]}|${r[1]}`, Number(r.cnt ?? r[2])]));

  const byModelList = Array.from(byModelMap.entries())
    .map(([key, stats]) => ({ ...stats, conversations: convMap.get(key) ?? 0 }))
    .sort((a, b) => b.prompts - a.prompts);

  // By provider
  const provRows = (await client.execute({
    sql: `SELECT COALESCE(model_provider, 'unknown'), platform, COUNT(*) as cnt, SUM(word_count) as words FROM messages WHERE role = 'human' AND (model_id IS NOT NULL OR model_provider IS NOT NULL)${pf.clause} GROUP BY 1, platform`,
    args: pf.args,
  })).rows;

  const byProvMap = new Map<string, any>();
  for (const r of provRows) {
    const prov = String(r[0]);
    if (!byProvMap.has(prov)) byProvMap.set(prov, { model_provider: prov, prompts: 0, words: 0, sources: {} });
    const entry = byProvMap.get(prov)!;
    entry.prompts += Number(r.cnt ?? r[2]);
    entry.words += Number(r.words ?? r[3]);
    entry.sources[String(r.platform ?? r[1])] = Number(r.cnt ?? r[2]);
  }

  const provConvRows = (await client.execute({
    sql: `SELECT COALESCE(model_provider, 'unknown'), COUNT(DISTINCT conversation_id) as cnt FROM messages WHERE role = 'human' AND (model_id IS NOT NULL OR model_provider IS NOT NULL)${pf.clause} GROUP BY 1`,
    args: pf.args,
  })).rows;
  const provConvMap = new Map(provConvRows.map((r) => [String(r[0]), Number(r.cnt ?? r[1])]));

  const byProvList = Array.from(byProvMap.entries())
    .map(([prov, stats]) => ({ ...stats, conversations: provConvMap.get(prov) ?? 0 }))
    .sort((a, b) => b.prompts - a.prompts);

  // Time series
  const tsRows = (await client.execute({
    sql: `SELECT platform, local_date, COALESCE(model_id, 'unknown') as mid, COUNT(*) as cnt FROM messages WHERE role = 'human' AND (model_id IS NOT NULL OR model_provider IS NOT NULL)${pf.clause} GROUP BY platform, local_date, 3 ORDER BY platform, local_date`,
    args: pf.args,
  })).rows;

  const bySourceDate = new Map<string, Map<string, Record<string, number>>>();
  for (const r of tsRows) {
    const plat = String(r.platform ?? r[0]);
    const date = String(r.local_date ?? r[1]);
    const mid = String(r.mid ?? r[2]);
    const cnt = Number(r.cnt ?? r[3]);
    if (!bySourceDate.has(plat)) bySourceDate.set(plat, new Map());
    const dayMap = bySourceDate.get(plat)!;
    if (!dayMap.has(date)) dayMap.set(date, {});
    dayMap.get(date)![mid] = cnt;
  }

  const timeSeriesBySource: Record<string, any[]> = {};
  for (const [source, dayMap] of bySourceDate) {
    timeSeriesBySource[source] = Array.from(dayMap.entries())
      .sort(([a], [b]) => a.localeCompare(b))
      .map(([date, models]) => ({
        date,
        total_prompts: Object.values(models).reduce((a, b) => a + b, 0),
        models: Object.fromEntries(Object.entries(models).sort(([, a], [, b]) => b - a)),
      }));
  }

  return {
    coverage: { total_human_prompts: totalHuman, prompts_with_model_metadata: withMeta, metadata_coverage_pct: coveragePct },
    by_model: byModelList,
    by_provider: byProvList,
    time_series_by_source: timeSeriesBySource,
  };
}

export async function computeMetrics(
  client: Client,
  config: Config,
  platform?: Platform,
): Promise<Record<string, any>> {
  const pf = platformFilter(platform);

  // Volume
  const vol = (await client.execute({
    sql: `SELECT COUNT(*) as total, SUM(CASE WHEN role = 'human' THEN 1 ELSE 0 END) as human, SUM(CASE WHEN role = 'assistant' THEN 1 ELSE 0 END) as assistant, SUM(CASE WHEN role = 'human' THEN word_count ELSE 0 END) as words_h, SUM(CASE WHEN role = 'assistant' THEN word_count ELSE 0 END) as words_a, COUNT(DISTINCT conversation_id) as convos FROM messages WHERE 1=1${pf.clause}`,
    args: pf.args,
  })).rows[0];

  const totalMessages = Number(vol.total);
  const totalHuman = Number(vol.human);
  const totalAssistant = Number(vol.assistant);
  const totalWordsHuman = Number(vol.words_h ?? 0);
  const totalWordsAssistant = Number(vol.words_a ?? 0);
  const totalConversations = Number(vol.convos);

  if (!totalHuman) throw new Error("No human messages found in data");
  const avgWordsPerPrompt = totalWordsHuman / totalHuman;

  // Conversation depth
  const depthRows = (await client.execute({
    sql: `SELECT conversation_id, COUNT(*) as cnt FROM messages WHERE 1=1${pf.clause} GROUP BY conversation_id`,
    args: pf.args,
  })).rows;
  const turns = depthRows.map((r) => Number(r.cnt));
  const avgTurns = turns.length ? turns.reduce((a, b) => a + b, 0) / turns.length : 0;
  const maxTurns = turns.length ? Math.max(...turns) : 0;
  const quickAsks = turns.filter((t) => t <= 3).length;
  const workingSessions = turns.filter((t) => t >= 4 && t <= 10).length;
  const deepDives = turns.filter((t) => t > 10).length;
  const responseRatio = totalWordsHuman ? totalWordsAssistant / totalWordsHuman : 0;

  // Temporal
  const heatmapRows = (await client.execute({
    sql: `SELECT local_weekday, local_hour, COUNT(*) as cnt FROM messages WHERE role = 'human'${pf.clause} GROUP BY local_weekday, local_hour`,
    args: pf.args,
  })).rows;
  const heatmap: number[][] = Array.from({ length: 7 }, () => Array(24).fill(0));
  for (const r of heatmapRows) heatmap[Number(r.local_weekday)][Number(r.local_hour)] = Number(r.cnt);

  const nightOwl = Number((await client.execute({
    sql: `SELECT COUNT(*) as cnt FROM messages WHERE role = 'human' AND (local_hour >= 23 OR local_hour < 4)${pf.clause}`,
    args: pf.args,
  })).rows[0].cnt);
  const nightOwlPct = (nightOwl / totalHuman) * 100;

  const hourRows = (await client.execute({
    sql: `SELECT local_hour, COUNT(*) as cnt FROM messages WHERE role = 'human'${pf.clause} GROUP BY local_hour`,
    args: pf.args,
  })).rows;
  const hourCounts: Record<number, number> = {};
  let peakHour = 0, peakHourCount = 0;
  for (const r of hourRows) {
    const h = Number(r.local_hour);
    const c = Number(r.cnt);
    hourCounts[h] = c;
    if (c > peakHourCount) { peakHour = h; peakHourCount = c; }
  }

  const dayRows = (await client.execute({
    sql: `SELECT local_weekday, COUNT(*) as cnt FROM messages WHERE role = 'human'${pf.clause} GROUP BY local_weekday`,
    args: pf.args,
  })).rows;
  let peakDay = 0, peakDayCount = 0;
  for (const r of dayRows) {
    const d = Number(r.local_weekday);
    const c = Number(r.cnt);
    if (c > peakDayCount) { peakDay = d; peakDayCount = c; }
  }
  const dayNames = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"];

  // Style metrics (regex on content)
  const humanMsgs = await queryMessages(client, { role: Role.HUMAN, platform });
  const assistantMsgs = await queryMessages(client, { role: Role.ASSISTANT, platform });
  const humanTexts = humanMsgs.map((m) => m.content);
  const assistantTexts = assistantMsgs.map((m) => m.content);

  const polCounts: Record<string, number> = {};
  for (const [word, pattern] of Object.entries(POLITENESS_PATTERNS)) {
    polCounts[word] = countAllMatches(humanTexts, pattern);
  }
  const totalPoliteness = Object.values(polCounts).reduce((a, b) => a + b, 0);
  const polPer100 = (totalPoliteness / totalHuman) * 100;

  const btCounts: Record<string, number> = {};
  for (const [word, pattern] of Object.entries(BACKTRACK_PATTERNS)) {
    btCounts[word] = countAllMatches(humanTexts, pattern);
  }
  const totalBacktrack = Object.values(btCounts).reduce((a, b) => a + b, 0);
  const btPer100 = (totalBacktrack / totalHuman) * 100;

  const qCount = humanMsgs.filter((m) => m.content.trim().endsWith("?")).length;
  const qRate = (qCount / totalHuman) * 100;
  let cCount = 0;
  for (const m of humanMsgs) {
    if (COMMAND_PATTERN.test(m.content.trim())) cCount++;
    COMMAND_PATTERN.lastIndex = 0;
  }
  const cRate = (cCount / totalHuman) * 100;

  const yrPattern = /you'?re (absolutely )?right/gi;
  const yrCount = countAllMatches(assistantTexts, yrPattern);
  const yrPerConvo = totalConversations ? yrCount / totalConversations : 0;

  // Platform stats
  const platRows = (await client.execute({
    sql: `SELECT platform, COUNT(*) as cnt, SUM(word_count) as words, COUNT(DISTINCT conversation_id) as convos, MIN(timestamp) as first_ts FROM messages WHERE role = 'human'${pf.clause} GROUP BY platform`,
    args: pf.args,
  })).rows;
  const platformStats: Record<string, any> = {};
  for (const r of platRows) {
    platformStats[String(r.platform)] = {
      messages: Number(r.cnt),
      words: Number(r.words),
      conversations: Number(r.convos),
      first_message: String(r.first_ts),
    };
  }

  // Persona
  const persona = classifyPersona(polPer100, btPer100, qRate, cRate, config);

  // Date range
  const dateRange = (await client.execute({
    sql: `SELECT MIN(timestamp) as first_ts, MAX(timestamp) as last_ts FROM messages WHERE role = 'human'${pf.clause}`,
    args: pf.args,
  })).rows[0];

  // Sub-module metrics
  const [modelUsage, trends, nlp] = await Promise.all([
    computeModelUsage(client, platform),
    computeTrendMetrics(client, platform),
    computeNlpMetrics(client, platform),
  ]);

  return {
    generated_at: new Date().toISOString(),
    volume: {
      total_messages: totalMessages,
      total_human: totalHuman,
      total_assistant: totalAssistant,
      total_words_human: totalWordsHuman,
      total_words_assistant: totalWordsAssistant,
      total_conversations: totalConversations,
      avg_words_per_prompt: round(avgWordsPerPrompt, 1),
    },
    conversation_depth: {
      avg_turns: round(avgTurns, 1),
      max_turns: maxTurns,
      quick_asks: quickAsks,
      working_sessions: workingSessions,
      deep_dives: deepDives,
    },
    response_ratio: round(responseRatio, 1),
    temporal: {
      heatmap,
      night_owl_pct: round(nightOwlPct, 1),
      peak_hour: peakHour,
      peak_hour_count: peakHourCount,
      peak_day: dayNames[peakDay],
      peak_day_count: peakDayCount,
      hour_counts: hourCounts,
    },
    politeness: { counts: polCounts, total: totalPoliteness, per_100_prompts: round(polPer100, 1) },
    backtrack: { counts: btCounts, total: totalBacktrack, per_100_prompts: round(btPer100, 1) },
    question: { count: qCount, rate: round(qRate, 1) },
    command: { count: cCount, rate: round(cRate, 1) },
    youre_right: { count: yrCount, per_conversation: round(yrPerConvo, 1) },
    persona,
    platform_stats: platformStats,
    model_usage: modelUsage,
    trends,
    nlp,
    date_range: { first: String(dateRange.first_ts), last: String(dateRange.last_ts) },
  };
}

export async function hasHumanMessages(client: Client, platform?: Platform): Promise<boolean> {
  const pf = platformFilter(platform);
  const r = await client.execute({ sql: `SELECT COUNT(*) as cnt FROM messages WHERE role = 'human'${pf.clause}`, args: pf.args });
  return Number(r.rows[0].cnt) > 0;
}

export async function computeSourceViews(
  client: Client,
  config: Config,
): Promise<{ sourceViews: Record<string, any>; metadata: { default_view: string } }> {
  const allMetrics = await computeMetrics(client, config);
  const ccHas = await hasHumanMessages(client, Platform.CLAUDE_CODE);
  const cxHas = await hasHumanMessages(client, Platform.CODEX);
  const agHas = await hasHumanMessages(client, Platform.AGENT);

  const sourceViews: Record<string, any> = {
    both: allMetrics,
    claude_code: ccHas ? await computeMetrics(client, config, Platform.CLAUDE_CODE) : null,
    codex: cxHas ? await computeMetrics(client, config, Platform.CODEX) : null,
    agent: agHas ? await computeMetrics(client, config, Platform.AGENT) : null,
  };

  let defaultView = "both";
  if (!sourceViews[defaultView]) {
    for (const candidate of ["claude_code", "codex"]) {
      if (sourceViews[candidate]) { defaultView = candidate; break; }
    }
  }

  return { sourceViews, metadata: { default_view: defaultView } };
}

File: src-py/metrics.py
"""Metric computation: volume, depth, temporal, style, and source views."""

import re
import sqlite3
from collections import defaultdict
from datetime import datetime, timezone

from .config import Config
from .db import platform_filter, query_messages
from .models import Message, Platform, Role
from .nlp import BACKTRACK_PATTERNS, COMMAND_PATTERN, POLITENESS_PATTERNS, compute_nlp_metrics
from .persona import classify_persona
from .trends import compute_trend_metrics


def compute_model_usage(conn: sqlite3.Connection, platform: Platform | None = None) -> dict:
    """Aggregate model usage across source and time for human prompts."""
    pf, pp = platform_filter(platform)

    # Coverage
    total_human = conn.execute(
        f"SELECT COUNT(*) FROM messages WHERE role = 'human'{pf}", pp,
    ).fetchone()[0]
    with_metadata = conn.execute(
        f"SELECT COUNT(*) FROM messages"
        f" WHERE role = 'human' AND (model_id IS NOT NULL OR model_provider IS NOT NULL){pf}",
        pp,
    ).fetchone()[0]
    coverage_pct = round((with_metadata / total_human * 100), 1) if total_human else 0.0

    # --- By model (with per-platform source breakdown) ---
    model_source_rows = conn.execute(
        f"SELECT COALESCE(model_id, 'unknown'), COALESCE(model_provider, 'unknown'),"
        f" platform, COUNT(*), SUM(word_count)"
        f" FROM messages"
        f" WHERE role = 'human' AND (model_id IS NOT NULL OR model_provider IS NOT NULL){pf}"
        f" GROUP BY 1, 2, platform",
        pp,
    ).fetchall()

    by_model_map: dict[tuple, dict] = {}
    for mid, mprov, plat, prompts, words in model_source_rows:
        key = (mid, mprov)
        if key not in by_model_map:
            by_model_map[key] = {
                "model_id": mid, "model_provider": mprov,
                "prompts": 0, "words": 0, "sources": {},
            }
        by_model_map[key]["prompts"] += prompts
        by_model_map[key]["words"] += words
        by_model_map[key]["sources"][plat] = prompts

    # Conversation counts per model
    conv_rows = conn.execute(
        f"SELECT COALESCE(model_id, 'unknown'), COALESCE(model_provider, 'unknown'),"
        f" COUNT(DISTINCT conversation_id)"
        f" FROM messages"
        f" WHERE role = 'human' AND (model_id IS NOT NULL OR model_provider IS NOT NULL){pf}"
        f" GROUP BY 1, 2",
        pp,
    ).fetchall()
    conv_map = {(r[0], r[1]): r[2] for r in conv_rows}

    by_model_list = [
        {
            "model_id": stats["model_id"],
            "model_provider": stats["model_provider"],
            "prompts": stats["prompts"],
            "words": stats["words"],
            "conversations": conv_map.get(key, 0),
            "sources": dict(sorted(stats["sources"].items())),
        }
        for key, stats in by_model_map.items()
    ]
    by_model_list.sort(key=lambda x: (-x["prompts"], x["model_id"], x["model_provider"]))

    # --- By provider ---
    prov_source_rows = conn.execute(
        f"SELECT COALESCE(model_provider, 'unknown'), platform,"
        f" COUNT(*), SUM(word_count)"
        f" FROM messages"
        f" WHERE role = 'human' AND (model_id IS NOT NULL OR model_provider IS NOT NULL){pf}"
        f" GROUP BY 1, platform",
        pp,
    ).fetchall()

    by_provider_map: dict[str, dict] = {}
    for mprov, plat, prompts, words in prov_source_rows:
        if mprov not in by_provider_map:
            by_provider_map[mprov] = {"model_provider": mprov, "prompts": 0, "words": 0, "sources": {}}
        by_provider_map[mprov]["prompts"] += prompts
        by_provider_map[mprov]["words"] += words
        by_provider_map[mprov]["sources"][plat] = prompts

    prov_conv_rows = conn.execute(
        f"SELECT COALESCE(model_provider, 'unknown'), COUNT(DISTINCT conversation_id)"
        f" FROM messages"
        f" WHERE role = 'human' AND (model_id IS NOT NULL OR model_provider IS NOT NULL){pf}"
        f" GROUP BY 1",
        pp,
    ).fetchall()
    prov_conv_map = {r[0]: r[1] for r in prov_conv_rows}

    by_provider_list = [
        {
            "model_provider": stats["model_provider"],
            "prompts": stats["prompts"],
            "words": stats["words"],
            "conversations": prov_conv_map.get(mprov, 0),
            "sources": dict(sorted(stats["sources"].items())),
        }
        for mprov, stats in by_provider_map.items()
    ]
    by_provider_list.sort(key=lambda x: (-x["prompts"], x["model_provider"]))

    # --- Time series by source ---
    ts_rows = conn.execute(
        f"SELECT platform, local_date, COALESCE(model_id, 'unknown'), COUNT(*)"
        f" FROM messages"
        f" WHERE role = 'human' AND (model_id IS NOT NULL OR model_provider IS NOT NULL){pf}"
        f" GROUP BY platform, local_date, 3"
        f" ORDER BY platform, local_date",
        pp,
    ).fetchall()

    by_source_date: dict[str, dict[str, dict[str, int]]] = defaultdict(lambda: defaultdict(dict))
    for plat, date, mid, cnt in ts_rows:
        by_source_date[plat][date][mid] = cnt

    time_series_by_source = {}
    for source, day_map in by_source_date.items():
        source_series = []
        for date_key in sorted(day_map.keys()):
            model_counts = day_map[date_key]
            source_series.append({
                "date": date_key,
                "total_prompts": sum(model_counts.values()),
                "models": dict(sorted(model_counts.items(), key=lambda kv: (-kv[1], kv[0]))),
            })
        time_series_by_source[source] = source_series

    return {
        "coverage": {
            "total_human_prompts": total_human,
            "prompts_with_model_metadata": with_metadata,
            "metadata_coverage_pct": coverage_pct,
        },
        "by_model": by_model_list,
        "by_provider": by_provider_list,
        "time_series_by_source": time_series_by_source,
    }


def compute_metrics(conn: sqlite3.Connection, config: Config, platform: Platform | None = None) -> dict:
    """Compute all metrics from the database."""
    pf, pp = platform_filter(platform)

    # === Volume Metrics (SQL) ===
    vol = conn.execute(
        f"SELECT"
        f" COUNT(*),"
        f" SUM(CASE WHEN role = 'human' THEN 1 ELSE 0 END),"
        f" SUM(CASE WHEN role = 'assistant' THEN 1 ELSE 0 END),"
        f" SUM(CASE WHEN role = 'human' THEN word_count ELSE 0 END),"
        f" SUM(CASE WHEN role = 'assistant' THEN word_count ELSE 0 END),"
        f" COUNT(DISTINCT conversation_id)"
        f" FROM messages WHERE 1=1{pf}",
        pp,
    ).fetchone()

    total_messages = vol[0]
    total_human = vol[1]
    total_assistant = vol[2]
    total_words_human = vol[3] or 0
    total_words_assistant = vol[4] or 0
    total_conversations = vol[5]

    if not total_human:
        raise ValueError("No human messages found in data")

    avg_words_per_prompt = total_words_human / total_human

    # === Conversation Depth (SQL) ===
    depth_rows = conn.execute(
        f"SELECT conversation_id, COUNT(*) FROM messages WHERE 1=1{pf} GROUP BY conversation_id",
        pp,
    ).fetchall()

    turns_values = [r[1] for r in depth_rows]
    avg_turns = sum(turns_values) / len(turns_values) if turns_values else 0
    max_turns = max(turns_values) if turns_values else 0
    quick_asks = sum(1 for t in turns_values if t <= 3)
    working_sessions = sum(1 for t in turns_values if 4 <= t <= 10)
    deep_dives = sum(1 for t in turns_values if t > 10)
    response_ratio = total_words_assistant / total_words_human if total_words_human else 0

    # === Temporal Metrics (SQL via pre-computed local_hour/local_weekday) ===
    heatmap_rows = conn.execute(
        f"SELECT local_weekday, local_hour, COUNT(*)"
        f" FROM messages WHERE role = 'human'{pf}"
        f" GROUP BY local_weekday, local_hour",
        pp,
    ).fetchall()

    heatmap = defaultdict(lambda: defaultdict(int))
    for dow, hour, count in heatmap_rows:
        heatmap[dow][hour] = count
    heatmap_data = [[heatmap[dow][hour] for hour in range(24)] for dow in range(7)]

    night_owl_count = conn.execute(
        f"SELECT COUNT(*) FROM messages"
        f" WHERE role = 'human' AND (local_hour >= 23 OR local_hour < 4){pf}",
        pp,
    ).fetchone()[0]
    night_owl_pct = (night_owl_count / total_human * 100)

    hour_rows = conn.execute(
        f"SELECT local_hour, COUNT(*) as cnt"
        f" FROM messages WHERE role = 'human'{pf}"
        f" GROUP BY local_hour",
        pp,
    ).fetchall()
    hour_counts = {r[0]: r[1] for r in hour_rows}
    peak_hour_row = max(hour_rows, key=lambda x: x[1]) if hour_rows else (0, 0)
    peak_hour = peak_hour_row[0]
    peak_hour_count = peak_hour_row[1]

    day_rows = conn.execute(
        f"SELECT local_weekday, COUNT(*) as cnt"
        f" FROM messages WHERE role = 'human'{pf}"
        f" GROUP BY local_weekday",
        pp,
    ).fetchall()
    peak_day_row = max(day_rows, key=lambda x: x[1]) if day_rows else (0, 0)
    peak_day = peak_day_row[0]
    peak_day_count = peak_day_row[1]
    day_names = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']

    # === Style Metrics (Python regex on DB content) ===
    human_msgs = query_messages(conn, role=Role.HUMAN, platform=platform)
    assistant_msgs = query_messages(conn, role=Role.ASSISTANT, platform=platform)

    # Politeness
    politeness_counts = {}
    for word, pattern in POLITENESS_PATTERNS.items():
        count = sum(len(re.findall(pattern, m.content, re.IGNORECASE)) for m in human_msgs)
        politeness_counts[word] = count
    total_politeness = sum(politeness_counts.values())
    politeness_per_100 = (total_politeness / total_human * 100)

    # Backtrack
    backtrack_counts = {}
    for word, pattern in BACKTRACK_PATTERNS.items():
        count = sum(len(re.findall(pattern, m.content, re.IGNORECASE)) for m in human_msgs)
        backtrack_counts[word] = count
    total_backtrack = sum(backtrack_counts.values())
    backtrack_per_100 = (total_backtrack / total_human * 100)

    # Question / Command rates
    question_count = sum(1 for m in human_msgs if m.content.strip().endswith('?'))
    question_rate = (question_count / total_human * 100)
    command_count = sum(
        1 for m in human_msgs if re.match(COMMAND_PATTERN, m.content.strip(), re.IGNORECASE)
    )
    command_rate = (command_count / total_human * 100)

    # "You're absolutely right" count
    youre_right_count = sum(
        len(re.findall(r"you'?re (absolutely )?right", m.content, re.IGNORECASE))
        for m in assistant_msgs
    )
    youre_right_per_convo = youre_right_count / total_conversations if total_conversations else 0

    # === Platform Stats (SQL) ===
    plat_rows = conn.execute(
        f"SELECT platform, COUNT(*), SUM(word_count),"
        f" COUNT(DISTINCT conversation_id), MIN(timestamp)"
        f" FROM messages WHERE role = 'human'{pf}"
        f" GROUP BY platform",
        pp,
    ).fetchall()

    platform_stats = {}
    for plat, msgs, words, convos, first_ts in plat_rows:
        platform_stats[plat] = {
            "messages": msgs,
            "words": words,
            "conversations": convos,
            "first_message": first_ts,
        }

    # === Persona Classification ===
    persona = classify_persona(politeness_per_100, backtrack_per_100, question_rate, command_rate, config)

    # === Date Range (SQL) ===
    date_range_row = conn.execute(
        f"SELECT MIN(timestamp), MAX(timestamp) FROM messages WHERE role = 'human'{pf}",
        pp,
    ).fetchone()

    # === Sub-module metrics ===
    model_usage = compute_model_usage(conn, platform)
    trends = compute_trend_metrics(conn, platform)
    nlp = compute_nlp_metrics(conn, platform)

    return {
        "generated_at": datetime.now(timezone.utc).isoformat(),
        "volume": {
            "total_messages": total_messages,
            "total_human": total_human,
            "total_assistant": total_assistant,
            "total_words_human": total_words_human,
            "total_words_assistant": total_words_assistant,
            "total_conversations": total_conversations,
            "avg_words_per_prompt": round(avg_words_per_prompt, 1),
        },
        "conversation_depth": {
            "avg_turns": round(avg_turns, 1),
            "max_turns": max_turns,
            "quick_asks": quick_asks,
            "working_sessions": working_sessions,
            "deep_dives": deep_dives,
        },
        "response_ratio": round(response_ratio, 1),
        "temporal": {
            "heatmap": heatmap_data,
            "night_owl_pct": round(night_owl_pct, 1),
            "peak_hour": peak_hour,
            "peak_hour_count": peak_hour_count,
            "peak_day": day_names[peak_day],
            "peak_day_count": peak_day_count,
            "hour_counts": dict(hour_counts),
        },
        "politeness": {
            "counts": politeness_counts,
            "total": total_politeness,
            "per_100_prompts": round(politeness_per_100, 1),
        },
        "backtrack": {
            "counts": backtrack_counts,
            "total": total_backtrack,
            "per_100_prompts": round(backtrack_per_100, 1),
        },
        "question": {
            "count": question_count,
            "rate": round(question_rate, 1),
        },
        "command": {
            "count": command_count,
            "rate": round(command_rate, 1),
        },
        "youre_right": {
            "count": youre_right_count,
            "per_conversation": round(youre_right_per_convo, 1),
        },
        "persona": persona,
        "platform_stats": platform_stats,
        "model_usage": model_usage,
        "trends": trends,
        "nlp": nlp,
        "date_range": {
            "first": date_range_row[0],
            "last": date_range_row[1],
        },
    }


def has_human_messages_db(conn: sqlite3.Connection, platform: Platform | None = None) -> bool:
    """Return True if the DB contains at least one human message (optionally filtered by platform)."""
    pf, pp = platform_filter(platform)
    count = conn.execute(
        f"SELECT COUNT(*) FROM messages WHERE role = 'human'{pf}", pp,
    ).fetchone()[0]
    return count > 0


def compute_source_views(conn: sqlite3.Connection, config: Config) -> tuple[dict, dict]:
    """
    Build source-scoped metric views for dashboard filtering.

    Source semantics:
      - both: Claude Code + Codex
      - claude_code: Claude Code only
      - codex: Codex only
    """
    all_metrics = compute_metrics(conn, config)
    claude_code_metrics = (
        compute_metrics(conn, config, Platform.CLAUDE_CODE)
        if has_human_messages_db(conn, Platform.CLAUDE_CODE)
        else None
    )
    codex_metrics = (
        compute_metrics(conn, config, Platform.CODEX)
        if has_human_messages_db(conn, Platform.CODEX)
        else None
    )

    source_views = {
        "both": all_metrics,
        "claude_code": claude_code_metrics,
        "codex": codex_metrics,
    }

    default_view = "both"
    if source_views[default_view] is None:
        for candidate in ("claude_code", "codex"):
            if source_views[candidate] is not None:
                default_view = candidate
                break

    return source_views, {"default_view": default_view}
